src/crypto-providers/util.ts

```diff
@@ -0,0 +1,262 @@
+): BIP32Path | undefined => {
+  if (!signingFiles.length) return undefined
+  return signingFiles[i] ? signingFiles[i].path : signingFiles[0].path
+}
+
+const filterSigningFiles = (
+  signingFiles: HwSigningData[],
+): {paymentSigningFiles: HwSigningData[], stakeSigningFiles: HwSigningData[]} => {
+  const paymentSigningFiles = signingFiles.filter(
+    (signingFile) => signingFile.type === 0,
+  )
+  const stakeSigningFiles = signingFiles.filter(
+    (signingFile) => signingFile.type === 1,
+  )
+  return {
+    paymentSigningFiles,
+    stakeSigningFiles,
+  }
+}
+
+const findSigningPath = (
+  certPubKeyHash: Buffer, stakingSigningFiles: HwSigningData[],
+): BIP32Path | undefined => {
+  const signingFile = stakingSigningFiles.find((file) => {
+    const { pubKey } = XPubKey(file.cborXPubKeyHex)
+    const pubKeyHash = getPubKeyBlake2b224Hash(pubKey)
+    return !Buffer.compare(pubKeyHash, certPubKeyHash)
+  })
+  return signingFile?.path
+}
+
+const txHasPoolPoolRegistrationCert = (
+  certs: _Certificate[],
+): boolean => certs.some(
+  ({ type }) => type === TxCertificateKeys.STAKEPOOL_REGISTRATION,
+)
+
+const validateTx = (
+  txAux: _TxAux, paymentSigningFiles: HwSigningData[], stakeSigningFiles: HwSigningData[],
+): void => {
+  if (!txAux.inputs.length) throw NamedError('MissingInputError')
+  if (!txAux.outputs.length) throw NamedError('MissingOutputError')
+  if (paymentSigningFiles.length > txAux.inputs.length) {
+    throw NamedError('TooManySigningFilesError')
+  }
+  const requireStakingSigningFile = !!(txAux.certificates.length + txAux.withdrawals.length)
+  if (
+    requireStakingSigningFile && !stakeSigningFiles.length
+  ) throw NamedError('MissingStakingSigningFileError')
+}
+
+const validateWitnessing = (
+  txAux: _TxAux, signingFiles: HwSigningData[],
+): void => {
+  const {
+    paymentSigningFiles,
+    stakeSigningFiles,
+  } = filterSigningFiles(signingFiles)
+  validateTx(txAux, paymentSigningFiles, stakeSigningFiles)
+  if (!txHasPoolPoolRegistrationCert(txAux.certificates)) return
+
+  if (txAux.certificates.length !== 1) throw NamedError('MultipleCertificatesWithPoolRegError')
+  if (txAux.withdrawals.length) throw NamedError('WithdrawalIncludedWithPoolRegError')
+  if (paymentSigningFiles.length) throw NamedError('PaymentFileInlucedWithPoolRegError')
+  if (stakeSigningFiles.length !== 1) throw NamedError('MultipleStakingSigningFilesWithPoolRegError')
+}
+
+const validateSigning = (
+  txAux: _TxAux, signingFiles: HwSigningData[],
+): void => {
+  const {
+    paymentSigningFiles,
+    stakeSigningFiles,
+  } = filterSigningFiles(signingFiles)
+  if (txHasPoolPoolRegistrationCert(txAux.certificates)) throw NamedError('CantSignTxWithPoolReg')
+  validateTx(txAux, paymentSigningFiles, stakeSigningFiles)
+  if (!paymentSigningFiles.length) throw NamedError('MissingPaymentSigningFileError')
+}
+
+const _packBootStrapAddress = (
+  file: HwSigningData, network: Network,
+) => {
```

Comment:
should we add return types to these helpers?

Reply:
Fixed.